### PR TITLE
 Refactor TWP42: Bases de Datos using SQL.js 

### LINKED ISSUE #193: Refactor lecture TWP42: Bases de datos
This lecture has codeblocks for sqlite3 code instead of having an interactive activecode like other lectures. This is because Runestone does not support the sqlite3 library from python. Bython does not support it either.

Because of these reasons it is needed to find a way to use sqlite3 inside the book or replan this lecture in a way to teach with SQL.

By far there has been 3 attemps of refactoring the lecture, here they are order by closest to the solution:

1. Using the activecode SQL languge from Runestone: https://github.com/PyAr/PyZombis/pull/185
Some configuration to use this SQL tool from Runestone must be done, because when you try to access the file `surf.db`, a CORS error shows up in console and nothing is shown. There might be something related with the book not being served by Runestone because in this error it tries to access to the db from an URL in runestone site: `/runestone/books/published/PyZombis/_static/sql-wasm.wasm` asuming that the book is published.

2. Using sql.js library from javascript: https://github.com/PyAr/PyZombis/pull/184
This other option loads the database using the `.. raw:html` instruction from runestone.
This is the best attempt since you need to find a way to access to the global javascript variable `db` from the brython activecode block. Enter this PR for more detail, but if you call db in browser console you can see the database.

3. With indexedDB: https://github.com/PyAr/PyZombis/pull/171 
This aproach works, but the syntax changes a lot from the original lecture where `sqlite3` was used.

For example:
```
cursor.execute("select * from surfers  where age > 25")
linhas = cursor.fetchall()
for linha in linhas:
    print("id :",linha['id'])
    print("Nombre :",linha['name'])
    print("Pais  :",linha['country'])
    print("Media  :",linha['average'])
    print("Estilo :",linha['board'])
    print("Edad  :",linha['age'])
    print()
```
had to change to:
```
if(getattr(v, "edad") >=25):
    for key in ["nombre", "pais", "media", "estilo", "edad"]:
        print(f"{key}: {getattr(v, key)}")
  getattr(res, "continue")()
```

It is not a friendly syntax for new students that are learning this for the first time.

### PR DESCRIPTION
**Pull request by shivamshan**
Signed-off-by: Shivam Shandilya <shivam.stpaulsdarjeeling@gmail.com>

## Summary

This PR is based upon and fixes issues in PR #184, which ultimately may close #193.<br>
The issues in PR #184 (where the SQL.js database was not accessible through activecode block) were fixed by using the `document `module provided by activecode and some javascript manipulation.<br><br>

The following are the results of this PR:

- a sqlite3 wrapper was implemented to mimic its functionality since it is not supported by skulpt yet
- added rollback and commit functionality to wrapper
- any query (SELECT, INSERT, DELETE, etc.) can be enclosed in a string and can be executed
- a function was created to show data in tabular format
- the same format can be followed in other SQL lectures to showcase use of different queries
- end to end test using playwright and unittest was implemented

## Checklist

- [x] Variables, functions and comments are translated to Spanish
- [x] Functions follow underscore notation
- [x] Spell check done & typos fixed
- [x] All python code is PEP8 compliant
- [x] Test coverage with Playwright implemented; locators are Python code
- [x] Reviewers assigned (all peers & at least 1 mentor)

### SELECT query execution on existing table

d4fee712-0569-4574-966b-7104463da11b

### Creating new table and inserting data into it

faff906e-2283-4a05-991d-86c0bd9a2bc3

### Using the rollback statement
c66bf7f8-451f-4a53-aba5-409d53b27f6a

### Using the commit statement
f7233fa7-ea4c-47a1-9fd2-bf34bb301ad1

### CODE REVIEW

**Comment by shivamshan** [comment]
Hi @reingart @angelasofiaremolinagutierrez , should I also add unittests and coverage with playwright in this PR??

**Comment by reingart** [comment]
Great work @shivamshan !
There are some items that need to be revised, but in general it is useful as a Proof-of-Concept of the idea 
To move further, could you add an playwright end-2-end test? (see the /tests folder)
That will help you to get more familiar with the testing and CI tools

**Comment by shivamshan** [comment]
@reingart sure I will move forward with this. Also could you tell me what revisions need to be made?

**Review by reingart — CHANGES_REQUESTED** [review]
Please see the comments and add a e2e test with playwright

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 1]
```diff
@@ -1,44 +1,143 @@
...
 
 
+    En este ejercicio leeremos los datos de la tabla ``surfers`` y mostraremos
+    solo los mayores a 25 con una instrucción SQL (conocidas como query). 
+    Actualmente, la consulta devuelve todos los datos en la tabla ``surfers``.
+ 
+    ~~~~
+    import document, time, json
+ 
+    def formato_datos(data, width):
```
[the remark this answers is not in the document]
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst:86]: okay..but how would that wrapper be incorporated here in activecode? could you lead me to any example or resource

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 2]
```diff
@@ -1,44 +1,143 @@
...
+               db.run("INSERT INTO surfers (id, nombre, pais, media, estilo, edad) VALUES (102, \"Juan Martino\",\"Spain\",9.01,\"Gun\",36)");
+               db.run("INSERT INTO surfers (id, nombre, pais, media, estilo, edad) VALUES (103, \"Joseph 'smitty' Smyth\",\"USA\",8.85,\"Cruizer\",18)");
+               db.run("INSERT INTO surfers (id, nombre, pais, media, estilo, edad) VALUES (104, \"Stacey O'Neill\",\"Ireland\",8.91,\"Malibu\",22)");
+               db.run("INSERT INTO surfers (id, nombre, pais, media, estilo, edad) VALUES (105, \"Aideen 'board babe' Wu\",\"Japan\",8.65,\"Fish\",24)");
+               db.run("INSERT INTO surfers (id, nombre, pais, media, estilo, edad) VALUES (106, \"Zack 'bonnie-lad' MacFadden\",\"Scotland\",7.82,\"Thruster\",26)");
+               db.run("INSERT INTO surfers (id, nombre, pais, media, estilo, edad) VALUES (107, \"Aaron Valentino\",\"Italy\",8.98,\"Gun\",19)");
+
+               // writing table in browser console as test
+               res = db.exec(document.getElementById("create-db-script").getAttribute("query"));
```
[the remark this answers is not in the document]
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst:39]: it seems that brython code when run from the activecode block doesn't have DOM access, therefore the SQL.Database objects are not directly accessible...I will however look into the latter suggestion.

File: _sources/lectures/TWP42/TWP42_1.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst:115]: I have implemented a simple wrapper for sqlite3 that currently mimics connect, execute, fetchall and close functions. Although I have not removed the formato_datos function, I have moved it along with the hidden wrapper code so students need not worry about it. But they can still use it

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 3]
```diff
@@ -1,44 +1,143 @@
...
+    #modifique la consulta a continuación para mostrar los 
+    #detalles de las personas cuya edad es mayor de 25 años
+    query = """SELECT * FROM surfers"""
+ 
+ 
+    #usando el módulo de documento y SQL.js para ejecutar la 
+    #consulta en una base de datos predefinida
+    script_element = document.getElementById("create-db-script")
+    script_element.setAttribute("query", query)
+    time.sleep(1)
```
[the remark this answers is not in the document]
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst]: I have solved this issue by using a flag attribute that gets set only when the query gives a result. To handle bad joins, I have added a kind of timeout feature, that tells the user that the query has issues if the query takes longer that 5 seconds to give a result.

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 4]
```diff
@@ -1,44 +1,143 @@
...
+               
+               if(res !== undefined && res.length > 0 && res[0].values.length > 0) {
+                  document.getElementById("create-db-script").setAttribute("result",JSON.stringify(res));        
+                  console.log(JSON.stringify(res));
+               }
+
+            });
+      }
+
+
```
[the remark this answers is not in the document]
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst]: I have added suitable comments to increase readability.

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 5]
```diff
@@ -1,44 +1,143 @@
...
+      var observer = new MutationObserver(function(mutations) {
+      mutations.forEach(function(mutation) {
+         if (mutation.type === "attributes") {
+               getData(config);
+         }
+      });
+      });
+
```
[the remark this answers is not in the document]
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst]: I have added suitable comments to increase readability.

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 6]
```diff
@@ -1,44 +1,234 @@
...
+    cursor.execute(query)
 
+    # obtener los datos
+    linhas = cursor.fetchall()
 
-.. code-block :: python
+    # imprimir los datos
+    # Se ha predefinido la función formato_datos para imprimir el resultado en formato tabular
+    if(linhas!=None):
+       formato_datos(linhas, 30)
```
Comment (reingart) [inline _sources/lectures/TWP42/TWP42_1.rst:213]: Here please return something similar to the original example (a for iterating over a list of dicts is ok)
No need to format the output (it is great but hides a bit the details of DB-API concepts we want to teach)

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 7]
```diff
@@ -1,44 +1,234 @@
...
+    # el método de conexión de sqlite3 se ha importado para su uso
+    # uno puede hacer esto por la declaración
+    # from sqlite3 import connect
 
-+ ¡Usando una base de datos! En este caso, sqlite3
+    # inicie la conexión y conéctese a la tabla de surfers predefinida
+    base_datos = connect('surfers.db')
 
-+ Obs .: copie surfersDB.sdb a su directorio
+    # crear un cursor y ejecutar la consulta
```
Comment (reingart) [inline _sources/lectures/TWP42/TWP42_1.rst:201]: Please add the cursor row factory, so eventually the student could copy & paste the exercise and rut in a normal python interpreter
```
    base_datos.row_factory = sqlite3.Row
```
(you can comment it provisionally until the sqlite module is more mature)

**Review by reingart — CHANGES_REQUESTED** [review]
The changes look good!
I made some  minor comments
I tried the code of the next exercise and it seems the execute is not working correctly: (just prints `Please check your query`)
````
cursor.execute('''create table alumnos(login varchar(8),ra integer)''')
cursor.execute('insert into alumnos values("masanori",42)')
cursor.execute('insert into alumnos values("emengarda",666)')
cursor.execute('select * from alumnos')
````
http://pyar.github.io/PyZombis/200/lectures/TWP42/TWP42_2.html

Probably you need to :
 - [ ] review execute for DDL 
 - [ ] implement commit & rollback
 - [ ] enhance a bit the error reporting (raise the excepcion thrown by sqlite)

File: _sources/lectures/TWP42/TWP42_1.rst (hunk 6 above)
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst]: updated the function to format the result from sql.js t resemble sqlite3

File: _sources/lectures/TWP42/TWP42_1.rst (hunk 7 above)
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst:205]: added the above statement as comment

**Comment by shivamshan** [comment]
- I updated the execute function to resemble more like the sqlite3 code. Previously it was required that all the SQL statements be inside one string and that be passed as query. But now the user can have multiple execute sttements.
- I have added functions for rollback and commit functionality, but they are currently empty because commit and rollback feature  require persistent changes to the db which is not currently supported here. I am working on a way around.

> * [X]  review execute for DDL
> * [X]  implement commit & rollback
> * [X]  enhance a bit the error reporting (raise the excepcion thrown by sqlite)

**Review by reingart — CHANGES_REQUESTED** [review]
Great! 
I think there is a bug that is not cleaning the last query results (repeats the same data)
The second execute should be retriving rows from the new alumnos tables
```
base_datos = connect('surfers.db')
cursor = base_datos.cursor()
query = 'SELECT * FROM surfers'
cursor.execute(query)
for row in cursor.fetchall():
    print(row)

cursor.execute('''create table alumnos(login varchar(8),ra integer)''')
cursor.execute('insert into alumnos values("masanori",42)')
cursor.execute('insert into alumnos values("emengarda",666)')
cursor.execute('select * from alumnos')
rows = cursor.fetchall()
for row in cursor.fetchall():
    print(row)
```

**Comment by shivamshan** [comment]
yes, I was not clearing the previous query, therefore the final result comprised of results of both sets of queries...with only the first result getting parsed. Now, this issue is fixed.

**Review by reingart — CHANGES_REQUESTED** [review]
Perfect!
A final request: you should be giving to the user a more descriptive error.
For example, if I try to query a non-existing table by mistake: 
```
query = 'SELECT * FROM wrong_table'
```
It shows a generic error `Query error. Something is wrong with the query`
I have to open the JS console to see the detailed error message:
```
Error: no such table: wrong_table
    at c.handleError (sql-wasm.js:89:254)
    at c.exec (sql-wasm.js:87:6)
    at TWP42_1.html:372:22
```
Please raise a proper Python Exception with the SQLite error text `Error: no such table: wrong_table`, so the experience is more friendly to begginers

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 8]
```diff
@@ -1,44 +1,254 @@
...
+                return formato_datos(result)
+       
+       # the below code resets the flag and result attributes of the script tag for the next query
+       def close(self):
+          document.getElementById("create-db-script").setAttribute("result", "")
+          document.getElementById("create-db-script").setAttribute("flag", "0")               
+ 
+    class Connection:
+       def commit(self):
+          pass
```
Comment (reingart) [inline _sources/lectures/TWP42/TWP42_1.rst:194]: Please execute here a SQL "COMMIT" statement

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 9]
```diff
@@ -1,44 +1,254 @@
...
+       def rollback(self):
+          pass
```
Comment (reingart) [inline _sources/lectures/TWP42/TWP42_1.rst:196]: Please execute here a SQL "ROLLBACK" statement

**Review by reingart — CHANGES_REQUESTED** [review]

File: _sources/lectures/TWP42/TWP42_1.rst (hunk 9 above)
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst]: added rollback functionality

File: _sources/lectures/TWP42/TWP42_1.rst (hunk 8 above)
Comment (shivamshan) [reply _sources/lectures/TWP42/TWP42_1.rst]: added commit functionality but cannot view its functionality as database is not persistent and new db object is created with each fetchall statement

**Review by reingart — APPROVED** [review]

**Comment by shivamshan** [comment]
Hi @reingart , I have made some changes and added proof of concept for the rollback function. I have attached a gif showing that  both the rollback and commit statements work as they are supposed to. But they is a limitation that using the commit and rollback statements after a fetchall statement might not generate the correct output, the reason being that the db is not persistent and a new db object is created after every fetchall statment.

The following can be seen below, the commit functionality is only viable before a fetchall statement, as a new db object is created after:
68c39395-c5cf-4b49-8d09-c9b94cd0aaf6

The code to recreate this:
```
base_datos = connect('surfers.db')

cursor = base_datos.cursor()
cursor.execute('delete from surfers where edad<35')
base_datos.commit()
cursor.execute('SELECT * FROM surfers')

linhas = cursor.fetchall()
if(linhas!=None):
   for linha in linhas:
      print("ID    :",linha['id'])
      print("Nombre :",linha['nombre'])
      print("Pais  :",linha['pais'])
      print("Media  :",linha['media'])
      print("Estilo :",linha['estilo'])
      print("Edad  :",linha['edad'])
      print()

print("-----------------------------------------")
cursor.execute('SELECT * FROM surfers')
linhas = cursor.fetchall()
if(linhas!=None):
   for linha in linhas:
      print("ID    :",linha['id'])
      print("Nombre :",linha['nombre'])
      print("Pais  :",linha['pais'])
      print("Media  :",linha['media'])
      print("Estilo :",linha['estilo'])
      print("Edad  :",linha['edad'])
      print()
# cerrar el cursor
cursor.close()
```

**Comment by shivamshan** [comment]
UPDATE:-   
The previous problem has been fixed now. The commit functionality now resembles that of sqlite3. Although the db is still not persistent, I found a way around to get get results after commit stataements.

### Result
The difference in results can be seen. The results are now in accordance with the committed transactions.

1a3df712-0c6d-4b8b-a7e5-493ee371e761

The code to recreate this:- 
```
base_datos = connect('surfers.db')

cursor = base_datos.cursor()
cursor.execute('delete from surfers where edad<35')
base_datos.commit()
cursor.execute('SELECT * FROM surfers')

linhas = cursor.fetchall()
if(linhas!=None):
   for linha in linhas:
      print("ID    :",linha['id'])
      print("Nombre :",linha['nombre'])
      print("Pais  :",linha['pais'])
      print("Media  :",linha['media'])
      print("Estilo :",linha['estilo'])
      print("Edad  :",linha['edad'])
      print()

print("-----------------------------------------")
cursor.execute('SELECT * FROM surfers')
linhas = cursor.fetchall()
if(linhas!=None):
   for linha in linhas:
      print("ID    :",linha['id'])
      print("Nombre :",linha['nombre'])
      print("Pais  :",linha['pais'])
      print("Media  :",linha['media'])
      print("Estilo :",linha['estilo'])
      print("Edad  :",linha['edad'])
      print()
# cerrar el cursor
cursor.close()
```

File: _sources/lectures/TWP42/TWP42_1.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (reingart) [reply _sources/lectures/TWP42/TWP42_1.rst:53]: You could access  `browser.window.parent` in Brython if the name is declared in the global JS namespace
See [Objects in Javascript programs](https://www.brython.info/static_doc/en/jsobjects.html)
Activecode runs in a IFRAME, so you have to access the parent window.
For a PoC implementation see [a4052d6](https://github.com/PyAr/PyZombis/commit/a4052d6567e6e7239df2e05546ff5755c8179168).
Accessing directly to JS objects will help to remove the observer and DOM manipulation to pass data

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 10]
```diff
@@ -1,44 +1,277 @@
...
++ Esta tabla tiene la información de los participantes del torneo de surf organizada en las siguientes columnas: ``id``, ``nombre``, ``pais``, ``media``, ``estilo``, ``edad``.
++ ``id`` es un número entero que identifica al concursante.
++ ``nombre`` es una string con su nombre y su apodo (si tiene uno).
++ ``pais`` muestra el país que está representando en el torneo.
++ ``media`` es el promedio de puntos obtenidos en el torneo.
++ ``estilo`` es el estilo que tiene su tabla de surf.
++ ``edad`` es un número entero que muestra la edad del concursante.
+      
+
+.. activecode:: ac_l42_1
```
Comment (reingart) [inline _sources/lectures/TWP42/TWP42_1.rst:119]: Convert this activity from Skulp to Brython interpreter

File: _sources/lectures/TWP42/TWP42_1.rst [hunk 11]
```diff
@@ -1,44 +1,277 @@
...
+           Dict = {}
+           for c,d in zip(columns,v):
+               Dict[c] = d
+           final_list.append(Dict)
+       
+       return final_list
+ 
+    # creando un contenedor para imitar el módulo sqlite3
+    # actualmente solo algunas funciones como connect, execute, fetchall son imitados
+    class Cursor:
```
Comment (reingart) [inline _sources/lectures/TWP42/TWP42_1.rst:148]: Move The Cursor/Connection classes & connect function to a separate file `_static/sqlite.py`
Import it as in the Pygame ticket #201

**Review by reingart — CHANGES_REQUESTED** [review]
I was playing a bit more with SQL.js and made a PoC a4052d6567e6e7239df2e05546ff5755c8179168 with a more clean approach (creating and accessing the JS Database object directly from brython)
I left some comments so we can finally merge this:
 - [] Resolve the conflicts and merge the PoC
 - [ ] Switch from Skulpt to Brython interpreter
 - [ ] Use JS objects directly via window.parent
 - [ ] Move the code to `_static/sqlite.py` so it can be reusable in the next activity

It would be great if we could load the surfers.db from the `_static` folder, so the INSERTs can be skipped

**Comment by shivamshan** [comment]
> I was playing a bit more with SQL.js and made a PoC [a4052d6](https://github.com/PyAr/PyZombis/commit/a4052d6567e6e7239df2e05546ff5755c8179168) with a more clean approach (creating and accessing the JS Database object directly from brython) I left some comments so we can finally merge this:
> 
> * [x] Resolve the conflicts and merge the PoC
> * [x]  Switch from Skulpt to Brython interpreter
> * [x]  Use JS objects directly via window.parent
> * [x]  Move the code to `_static/sqlite.py` so it can be reusable in the next activity
> 
> It would be great if we could load the surfers.db from the `_static` folder, so the INSERTs can be skipped

**Review by reingart — APPROVED** [review]
Great! 
Please make another PR with a test using plywright (just executing the query and checking the result)

**Pull request merged**